File: weebasic.c
/****************************************************************************
*
* WEEBASIC
*
* Implementation of a parser and stack-based interpreter for a toy
* programming language. This language was built by Maxime Chevalier-Boisvert
* in 2021 as a study or interview, and is released under the unlicense.
*
****************************************************************************/

#include "stdbool.h"
#include "stdint.h"
#include "stdlib.h"
#include "stdio.h"
#include "string.h"
#include "ctype.h"
#include "assert.h"

// 64K instructions should be enough for anybody
#define MAX_INSTRS 65536
#define MAX_IDENT_LEN 64
#define MAX_LOCALS 128
#define MAX_STACK 32

// Kinds of instructions we support
typedef enum
{
    OP_EXIT = 0,
    OP_ERROR,
    OP_PUSH,
    OP_GETLOCAL,
    OP_SETLOCAL,
    OP_EQ,
    OP_LT,
    OP_IF,
    OP_IFNOT,
    OP_ADD,
    OP_SUB,
    OP_READINT,
    OP_PRINT
} opcode_t;

// Immutable, heap-allocated string object
typedef struct
{
    // String length, excluding null-terminator
    const size_t len;

    // String data
    const char* data;

} string_t;

typedef union
{
    uint64_t idx;
    int64_t int_val;
    string_t* str;

} value_t;

// Format of the instructions we implement
typedef struct
{
    opcode_t op;
    value_t imm;

} instr_t;

// Local variable declaration
typedef struct LocalVar
{
    // Name of the variable
    char ident[MAX_IDENT_LEN];

    // Index of the local variable
    size_t idx;

    // Next variable in the list
    struct LocalVar* next;

} local_t;

bool is_int(value_t val)
{
    return val.int_val & 1;
}

value_t tag(int64_t val)
{
    return (value_t)((val << 1) | 1);
}

int64_t untag(value_t val)
{
    assert (is_int(val));
    return val.int_val >> 1;
}

// Consume whitespace chars in the input
void eat_ws(char** pstr)
{
    while (true)
    {
        char ch = **pstr;

        switch (ch)
        {
            // Keep reading as long as we see whitespace
            case ' ':
            case '\t':
            case '\r':
            case '\n':
            break;

            // Not whitespace, stop
            default:
            return;
        }

        // Move to the next character
        (*pstr)++;
    }
}

// Consume single-line comments
void eat_comment(char** pstr)
{
    while (true)
    {
        char ch = **pstr;

        if (ch == '\n')
        {
            // Move to the next character
            (*pstr)++;
            break;
        }

        if (ch == '\0')
        {
            break;
        }

        // Move to the next character
        (*pstr)++;
    }
}

// Match a token in the input
bool match_token(char** pstr, const char* token)
{
    eat_ws(pstr);

    size_t len_tok = strlen(token);
    if (strncmp(*pstr, token, len_tok) == 0)
    {
        *pstr += len_tok;
        eat_ws(pstr);
        return true;
    }

    return false;
}

// Fail to parse if a given token is not there
void expect_token(char** pstr, const char* token)
{
    if (!match_token(pstr, token))
    {
        fprintf(stderr, "expected token \"%s\"\n", token);
        exit(-1);
    }
}

// Parse an identifier
void parse_ident(char** pstr, char* ident_out)
{
    size_t ident_len = 0;

    while (true)
    {
        char ch = **pstr;

        if (ident_len >= MAX_IDENT_LEN - 1)
        {
            fprintf(stderr, "identifier too long\n");
            exit(-1);
        }

        if (!isalnum(ch) && ch != '_')
        {
            break;
        }

        // Store this character
        ident_out[ident_len] = ch;
        ident_len++;

        // Move to the next character
        (*pstr)++;
    }

    if (ident_len == 0)
    {
        fprintf(stderr, "expected identifier\n");
        exit(-1);
    }

    assert(ident_len <= MAX_IDENT_LEN - 1);
    ident_out[ident_len] = '\0';
}

// Parse a positive integer constant
int64_t parse_int(char** pstr)
{
    int64_t num = 0;

    while (true)
    {
        char ch = **pstr;

        if (!isdigit(ch))
            break;

        // Store this digit
        int64_t digit = ch - '0';
        num = 10 * num + digit;

        // Move to the next character
        (*pstr)++;
    }

    return num;
}

// Try to find a local variable declaration
local_t* find_local(local_t* local_vars, const char* ident)
{
    for (local_t* var = local_vars; var != NULL; var = var->next)
    {
        if (strcmp(var->ident, ident) == 0)
        {
            return var;
        }
    }

    return NULL;
}

// Macros to append an instruction to the program
#define APPEND_INSN(op_type) ( insns[(*insn_idx)++] = (instr_t){ op_type }, &insns[(*insn_idx)-1] )
#define APPEND_INSN_IMM(op_type, imm) ( insns[(*insn_idx)++] = (instr_t){ op_type, imm }, &insns[(*insn_idx)-1] )

// Parse an atomic expression
void parse_atom(char** pstr, instr_t* insns, size_t* insn_idx, local_t* locals)
{
    char ch = **pstr;

    // Read an integer from the console
    if (match_token(pstr, "read_int"))
    {
        APPEND_INSN(OP_READINT);
        return;
    }

    // Integer constant
    if (isdigit(ch))
    {
        int64_t num = parse_int(pstr);
        APPEND_INSN_IMM(OP_PUSH, num);
        return;
    }

    // Reference to a variable
    if (isalpha(ch) || ch == '_')
    {
        // Parse the variable name
        char ident[MAX_IDENT_LEN];
        parse_ident(pstr, ident);

        // Try to find the declaration
        local_t* local = find_local(locals, ident);

        if (!local)
        {
            fprintf(stderr, "reference to undeclared variable \"%s\"\n", ident);
            exit(-1);
        }

        APPEND_INSN_IMM(OP_GETLOCAL, local->idx);
        return;
    }

    fprintf(stderr, "invalid expression");
    exit(-1);
}

// Parse an expression
void parse_expr(char** pstr, instr_t* insns, size_t* insn_idx, local_t* locals)
{
    // Parse a first expression
    parse_atom(pstr, insns, insn_idx, locals);

    eat_ws(pstr);

    char ch = **pstr;

    if (match_token(pstr, "+"))
    {
        // Parse the RHS expression
        parse_atom(pstr, insns, insn_idx, locals);

        // Add the result
        APPEND_INSN(OP_ADD);
        return;
    }

    if (match_token(pstr, "-"))
    {
        // Parse the RHS expression
        parse_atom(pstr, insns, insn_idx, locals);

        // Subtract the result
        APPEND_INSN(OP_SUB);
        return;
    }

    if (match_token(pstr, "=="))
    {
        // Parse the RHS expression
        parse_atom(pstr, insns, insn_idx, locals);

        // Compare the arguments
        APPEND_INSN(OP_EQ);
        return;
    }

    if (match_token(pstr, "<"))
    {
        // Parse the RHS expression
        parse_atom(pstr, insns, insn_idx, locals);

        // Compare the arguments
        APPEND_INSN(OP_LT);
        return;
    }
}

// Parse a statement
void parse_stmt(char** pstr, instr_t* insns, size_t* insn_idx, local_t** plocals)
{
    // Consume whitespace
    eat_ws(pstr);

    // Single-line comments
    if (match_token(pstr, "#"))
    {
        eat_comment(pstr);
        return;
    }

    // Local variable declaration
    if (match_token(pstr, "let"))
    {
        // Parse the variable name
        char ident[MAX_IDENT_LEN];
        parse_ident(pstr, ident);

        expect_token(pstr, "=");

        // Parse the expression we are assigning
        parse_expr(pstr, insns, insn_idx, *plocals);

        // Make sure this isn't a redeclaration
        local_t* first_local = *plocals;
        if (find_local(first_local, ident))
        {
            fprintf(stderr, "local variable \"%s\" already declared\n", ident);
            exit(-1);
        }

        // Create a new local variable
        local_t* new_local = malloc(sizeof(local_t));
        strcpy(new_local->ident, ident);
        new_local->idx = first_local? (first_local->idx + 1):0;
        new_local->next = first_local;
        *plocals = new_local;

        // Set the local to the expression's value
        APPEND_INSN_IMM(OP_SETLOCAL, new_local->idx);

        return;
    }

    if (match_token(pstr, "if"))
    {
        // Parse the test expression
        parse_expr(pstr, insns, insn_idx, *plocals);

        expect_token(pstr, "then");

        // If the result is false, jump past the if clause
        instr_t* ifnot_insn = APPEND_INSN_IMM(OP_IFNOT, 0);

        // Parse the body of the if statement
        parse_stmt(pstr, insns, insn_idx, plocals);

        // If the condition is false, we jump after the body of the if
        instr_t* pfalse = &insns[*insn_idx];
        ifnot_insn->imm.int_val = (pfalse - ifnot_insn) - 1;

        return;
    }

    // Sequencing of statements
    if (match_token(pstr, "begin"))
    {
        while (true)
        {
            if (match_token(pstr, "end"))
            {
                break;
            }

            parse_stmt(pstr, insns, insn_idx, plocals);
        }

        return;
    }

    // Print to stdout
    if (match_token(pstr, "print"))
    {
        parse_expr(pstr, insns, insn_idx, *plocals);
        APPEND_INSN(OP_PRINT);
        return;
    }

    // Print to stdout
    if (match_token(pstr, "assert"))
    {
        // Parse the condition
        parse_expr(pstr, insns, insn_idx, *plocals);

        // If the result is true, jump over the error instruction
        APPEND_INSN_IMM(OP_IF, 1);

        // Exit with an error
        APPEND_INSN(OP_ERROR);

        return;
    }

    // Cap the string length for printing
    if (strlen(*pstr) > 10)
    {
        (*pstr)[10] = '\0';
    }

    // Remove newlines from printout
    for (int i = 0;; ++i)
    {
        char ch = (*pstr)[i];
        if (ch == '\r' || ch == '\n')
            (*pstr)[i] = ' ';
        if (ch == '\0')
            break;
    }

    fprintf(stderr, "invalid statement: \"%s [...]\"\n", *pstr);
    exit(-1);
}

// Parse a source file into a sequence of instructions
instr_t* parse_file(const char* file_name)
{
    FILE* file = fopen(file_name, "r");

    if (!file)
    {
        fprintf(stderr, "failed to open source file \"%s\"\n", file_name);
        exit(-1);
    }

    // Read the entire file into a string buffer
    fseek(file, 0, SEEK_END);
    long fsize = ftell(file);
    fseek(file, 0, SEEK_SET);
    char* input_str = malloc(fsize + 1);
    fread(input_str, 1, fsize, file);
    input_str[fsize] = 0;
    fclose(file);

    // We will use a doubly-indirected pointer for parsing
    char* current_ch = input_str;
    char** pstr = &current_ch;

    // Instruction array
    instr_t* insns = malloc(sizeof(instr_t) * MAX_INSTRS);
    size_t insn_idx = 0;

    // Table of local variables
    local_t* local_vars = NULL;

    // Until we reach the end of the input
    while (true)
    {
        // End of input
        if (*current_ch == '\0')
        {
            break;
        }

        parse_stmt(pstr, insns, &insn_idx, &local_vars);
    }

    free(input_str);

    return insns;
}

// Stack manipulation primitives
#define PUSH(v) ( stack[stack_size] = v, stack_size++ )
#define POP() ( stack_size--, stack[stack_size] )

// Evaluate/run a program
void eval(const instr_t* insns)
{
    // Local variables
    value_t vars[MAX_LOCALS];

    // Stack of temporary values
    value_t stack[MAX_STACK];
    size_t stack_size = 0;

    for (const instr_t* pc = insns; pc != NULL; ++pc)
    {
        //printf("stack_size=%zu\n", stack_size);

        switch (pc->op)
        {
            // Exit the program
            case OP_EXIT:
            return;

            case OP_ERROR:
            fprintf(stderr, "Run-time error\n");
            exit(-1);
            return;

            case OP_PUSH:
            PUSH(pc->imm);
            break;

            case OP_SETLOCAL:
            vars[pc->imm.idx] = POP();
            break;

            case OP_GETLOCAL:
            PUSH(vars[pc->imm.idx]);
            break;

            case OP_EQ:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                int64_t bool_val = (arg0 == arg1)? 1:0;
                PUSH((value_t)bool_val);
            }
            break;

            case OP_LT:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                int64_t bool_val = (arg0 < arg1)? 1:0;
                PUSH((value_t)bool_val);
            }
            break;

            case OP_IF:
            {
                int64_t test_val = POP().int_val;

                if (test_val != 0)
                {
                    int64_t jump_offset = pc->imm.int_val;
                    pc += jump_offset;
                }
            }
            break;

            case OP_IFNOT:
            {
                int64_t test_val = POP().int_val;

                if (test_val == 0)
                {
                    int64_t jump_offset = pc->imm.int_val;
                    pc += jump_offset;
                }
            }
            break;

            case OP_ADD:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                PUSH((value_t)(arg0 + arg1));
            }
            break;

            case OP_SUB:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                PUSH((value_t)(arg0 - arg1));
            }
            break;

            case OP_READINT:
            {
                printf("Input an integer value:\n");
                printf("> ");

                int64_t int_val = 0;
                while (true)
                {
                    char c = fgetc(stdin);
                    if (c == EOF || !isdigit(c))
                        break;
                    int64_t digit = c - '0';
                    int_val = 10 * int_val + digit;
                }

                PUSH((value_t)int_val);
            }
            break;

            case OP_PRINT:
            {
                int64_t int_val = POP().int_val;
                printf("print: %lld\n", (long long)int_val);
            }
            break;

            default:
            fprintf(stderr, "unknown bytecode instruction\n");
            exit(-1);
        }
    }
}

int main(int argc, char** argv)
{
    if (argc == 2)
    {
        instr_t* insns = parse_file(argv[1]);
        eval(insns);
    }

    return 0;
}

File: weebasic.rs
/****************************************************************************
*
* WEEBASIC
*
* Implementation of a parser and stack-based interpreter for a toy
* programming language. This language was built by Maxime Chevalier-Boisvert
* in 2021-2022 for study or interview purposes, and is released under the
* unlicense.
*
****************************************************************************/

#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(dead_code)]
#![allow(unused_mut)]

use std::env;
use std::fs;
use std::collections::HashMap;

// Kinds of instructions (opcodes) we support
enum Op
{
    Exit,
    Error,
    Push,
    GetLocal,
    SetLocal,
    Equal,
    LessThan,
    IfTrue,
    IfNot,
    Add,
    Sub,
    ReadInt,
    Print
}

enum Value
{
    None,
    Idx(usize),
    IntVal(i64),
    Str(String),
}

/*
// Immutable, heap-allocated string object
typedef struct
{
    // String length, excluding null-terminator
    size_t len;

    // String data
    char* data;

} string_t;

typedef union
{
    uint64_t idx;
    int64_t int_val;
    string_t* str;

} value_t;

bool is_int(value_t val)
{
    return val.int_val & 1;
}

value_t tag(int64_t val)
{
    return (value_t)((val << 1) | 1);
}

int64_t untag(value_t val)
{
    assert (is_int(val));
    return val.int_val >> 1;
}
*/

// Format of the instructions we implement
struct Insn
{
    op: Op,
    imm: Value,
}

struct Program
{
    insns: Vec<Insn>,

    num_locals: usize,

    /// Mapping of identifiers to local variable indices
    local_idxs: HashMap<String, usize>,
}

impl Program
{
    fn new() -> Self
    {
        Program {
            insns: Vec::default(),
            num_locals: 0,
            local_idxs: HashMap::default(),
        }
    }

    /// Append an instruction with no argument
    fn append_insn(&mut self, op: Op)
    {
        self.insns.push(Insn {
            op: op,
            imm: Value::None
        });
    }

    /// Append an instruction with an immediate argument
    fn append_insn_imm(&mut self, op: Op, imm: Value)
    {
        self.insns.push(Insn {
            op: op,
            imm: imm
        });
    }

    /// Try to find the index for local variable declaration
    fn find_local(&self, ident: &str) -> Option<usize>
    {
        match self.local_idxs.get(ident) {
            Some(idx) => Some(*idx),
            None => None,
        }
    }

    /// Declare a new local variable
    fn declare_local(&mut self, ident: &str) -> usize
    {
        assert!(self.find_local(ident).is_none());
        let local_idx = self.local_idxs.len();
        self.local_idxs.insert(ident.to_owned(), local_idx);
        return local_idx;
    }
}

/// Stream of input characters to be parsed
struct Input
{
    /// Characters of the input string
    chars: Vec<char>,

    /// Current position in the input
    pos: usize,
}

impl Input
{
    fn new(input_str: String) -> Self
    {
        Input {
            chars: input_str.chars().collect(),
            pos: 0,
        }
    }

    /// Peek at the current input character
    fn peek_char(&self) -> char
    {
        if self.pos >= self.chars.len() {
            return '\0';
        }

        self.chars[self.pos]
    }

    /// Consume one character of the input
    fn eat_char(&mut self) -> char {
        let ch = self.peek_char();

        if ch != '\0' {
            self.pos += 1;
        }

        return ch
    }

    /// Consume whitespace chars in the input
    fn eat_ws(&mut self)
    {
        loop
        {
            let ch = self.peek_char();

            match ch
            {
                // Keep reading as long as we see whitespace
                ' ' | '\t' | '\r' | '\n' => {},

                // Not whitespace, stop
                _ => return,
            }

            // Move to the next character
            self.eat_char();
        }
    }

    /// Consume single-line comments
    fn eat_comment(&mut self)
    {
        loop
        {
            let ch = self.peek_char();

            match ch
            {
                '\n' => {
                    // Move to the next character
                    self.eat_char();
                    break;
                },

                '\0' => {
                    break;
                },

                _ => {
                    self.eat_char();
                }
            }
        }
    }

    /// Check if the input starts with a given token
    fn match_token(&mut self, token: &str) -> bool
    {
        self.eat_ws();

        let token_chars: Vec<char> = token.chars().collect();
        let num_chars = token_chars.len();

        if self.pos + num_chars > self.chars.len() {
            return false;
        }

        if self.chars[self.pos..(self.pos + num_chars)] == token_chars {
            self.pos += num_chars;
            self.eat_ws();
            return true;
        }

        return false;
    }

    /// Fail to parse if a given token is not there
    fn expect_token(&mut self, token: &str)
    {
        if !self.match_token(token) {
            panic!("expected token \"{}\"", token);
        }
    }

    /// Parse an identifier at the current position
    fn parse_ident(&mut self) -> String
    {
        let mut ident_str = String::from("");

        loop
        {
            let ch = self.peek_char();

            if !ch.is_alphanumeric() && ch != '_' {
                break;
            }

            // Store this character
            ident_str.push(ch);

            // Move to the next character
            self.eat_char();
        }

        if ident_str.len() == 0 {
            panic!("expected identifier\n");
        }

        return ident_str;
    }

    /// Parse a positive decimal integer constant
    fn parse_int(&mut self) -> i64
    {
        let mut num: i64 = 0;

        loop
        {
            let ch = self.peek_char();

            if !ch.is_digit(10) {
                break;
            }

            // Store this digit
            let digit: i64 = (ch as i64) - ('0' as i64);
            num = 10 * num + digit;

            // Move to the next character
            self.eat_char();
        }

        return num;
    }
}

// Parse an atomic expression
fn parse_atom(input: &mut Input, prog: &mut Program)
{
    let ch = input.peek_char();

    // Read an integer from the console
    if input.match_token("read_int") {
        prog.append_insn(Op::ReadInt);
        return;
    }

    // Integer constant
    if ch.is_digit(10) {
        let num = input.parse_int();
        prog.append_insn_imm(Op::Push, Value::IntVal(num));
        return;
    }

    // Reference to a variable
    if ch.is_alphabetic() || ch == '_' {
        // Parse the variable name
        let ident_str = input.parse_ident();

        // Try to find the declaration
        let local_idx = prog.find_local(&ident_str);

        if local_idx.is_none() {
            panic!("reference to undeclared variable \"{}\"\n", ident_str);
        }

        prog.append_insn_imm(Op::GetLocal, Value::Idx(local_idx.unwrap()));
        return;
    }

    panic!("invalid expression");
}

// Parse an expression
fn parse_expr(input: &mut Input, prog: &mut Program)
{
    // Parse a first expression
    parse_atom(input, prog);

    input.eat_ws();

    let ch = input.peek_char();

    if input.match_token("+") {
        // Parse the RHS expression
        parse_atom(input, prog);

        // Add the result
        prog.append_insn(Op::Add);
        return;
    }

    if input.match_token("-") {
        // Parse the RHS expression
        parse_atom(input, prog);

        // Subtract the result
        prog.append_insn(Op::Sub);
        return;
    }

    if input.match_token("==") {
        // Parse the RHS expression
        parse_atom(input, prog);

        // Compare the arguments
        prog.append_insn(Op::Equal);
        return;
    }

    if input.match_token("<") {
        // Parse the RHS expression
        parse_atom(input, prog);

        // Compare the arguments
        prog.append_insn(Op::LessThan);
        return;
    }
}

// Parse a statement
fn parse_stmt(input: &mut Input, prog: &mut Program)
{
    // Consume whitespace
    input.eat_ws();

    // Single-line comments
    if input.match_token("#") {
        input.eat_comment();
        return;
    }

    // Local variable declaration
    if input.match_token("let") {
        // Parse the variable name
        let ident_str = input.parse_ident();

        input.expect_token("=");

        // Parse the expression we are assigning
        parse_expr(input, prog);

        // Make sure this isn't a redeclaration
        let local_idx = prog.find_local(&ident_str);

        if local_idx.is_some() {
            panic!("local variable \"{}\" already declared\n", ident_str);
        }

        // Create a new local variable
        let local_idx = prog.declare_local(&ident_str);

        // Set the local to the expression's value
        prog.append_insn_imm(Op::SetLocal, Value::Idx(local_idx));

        return;
    }

    if input.match_token("if") {
        // Parse the test expression
        parse_expr(input, prog);

        input.expect_token("then");

        // If the result is false, jump past the if clause
        //instr_t* ifnot_insn = APPEND_INSN_IMM(OP_IFNOT, 0);
        let ifnot_insn_idx = prog.insns.len();
        prog.append_insn(Op::IfNot);

        // Parse the body of the if statement
        parse_stmt(input, prog);

        // If the condition is false, we jump after the body of the if
        let jumpto_idx = prog.insns.len();
        let jump_offset = (jumpto_idx as i64) - (ifnot_insn_idx as i64) - 1;
        prog.insns[ifnot_insn_idx].imm = Value::IntVal(jump_offset);

        return;
    }

    // Sequencing of statements
    if input.match_token("begin") {
        loop
        {
            if input.match_token("end") {
                break;
            }

            parse_stmt(input, prog);
        }

        return;
    }

    // Print to stdout
    if input.match_token("print") {
        parse_expr(input, prog);
        prog.append_insn(Op::Print);
        return;
    }

    // Assert that an expression evaluates to true
    if input.match_token("assert") {
        // Parse the condition
        parse_expr(input, prog);

        // If the result is true, jump over the error instruction
        prog.append_insn_imm(Op::IfTrue, Value::IntVal(1));

        // Exit with an error
        prog.append_insn(Op::Error);

        return;
    }

    /*
    // Cap the string length for printing
    if (strlen(*pstr) > 10)
    {
        (*pstr)[10] = '\0';
    }

    // Remove newlines from printout
    for (int i = 0;; ++i)
    {
        char ch = (*pstr)[i];
        if (ch == '\r' || ch == '\n')
            (*pstr)[i] = ' ';
        if (ch == '\0')
            break;
    }
    */

    panic!("invalid statement");
    //panic!("invalid statement: \"%s [...]\"\n", *pstr);
}

// Parse a source file into a sequence of instructions
fn parse_file(file_name: &str) -> Program
{
    let input_str = fs::read_to_string(file_name)
        .expect("couldn't read input source file");

    // Input to be parsed
    let mut input = Input::new(input_str);

    // Program being compiled
    let mut program: Program = Program::new();

    // Until we reach the end of the input
    loop
    {
        // End of input
        if input.peek_char() == '\0' {
            break;
        }

        parse_stmt(&mut input, &mut program);
    }

    return program;
}








/*
// Stack manipulation primitives
#define PUSH(v) ( stack[stack_size] = v, stack_size++ )
#define POP() ( stack_size--, stack[stack_size] )
*/

// Evaluate/run a program
fn eval(prog: Program)
{



    /*
    // Local variables
    value_t vars[MAX_LOCALS];

    // Stack of temporary values
    value_t stack[MAX_STACK];
    size_t stack_size = 0;

    for (const instr_t* pc = insns; pc != NULL; ++pc)
    {
        //printf("stack_size=%zu\n", stack_size);

        switch (pc->op)
        {
            // Exit the program
            case OP_EXIT:
            return;

            case OP_ERROR:
            fprintf(stderr, "Run-time error\n");
            exit(-1);
            return;

            case OP_PUSH:
            PUSH(pc->imm);
            break;

            case OP_SETLOCAL:
            vars[pc->imm.idx] = POP();
            break;

            case OP_GETLOCAL:
            PUSH(vars[pc->imm.idx]);
            break;

            case OP_EQ:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                int64_t bool_val = (arg0 == arg1)? 1:0;
                PUSH((value_t)bool_val);
            }
            break;

            case OP_LT:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                int64_t bool_val = (arg0 < arg1)? 1:0;
                PUSH((value_t)bool_val);
            }
            break;

            case OP_IF:
            {
                int64_t test_val = POP().int_val;

                if (test_val != 0)
                {
                    int64_t jump_offset = pc->imm.int_val;
                    pc += jump_offset;
                }
            }
            break;

            case OP_IFNOT:
            {
                int64_t test_val = POP().int_val;

                if (test_val == 0)
                {
                    int64_t jump_offset = pc->imm.int_val;
                    pc += jump_offset;
                }
            }
            break;

            case OP_ADD:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                PUSH((value_t)(arg0 + arg1));
            }
            break;

            case OP_SUB:
            {
                int64_t arg1 = POP().int_val;
                int64_t arg0 = POP().int_val;
                PUSH((value_t)(arg0 - arg1));
            }
            break;

            case OP_READINT:
            {
                printf("Input an integer value:\n");
                printf("> ");

                int64_t int_val = 0;
                while (true)
                {
                    char c = fgetc(stdin);
                    if (c == EOF || !isdigit(c))
                        break;
                    int64_t digit = c - '0';
                    int_val = 10 * int_val + digit;
                }

                PUSH((value_t)int_val);
            }
            break;

            case OP_PRINT:
            {
                int64_t int_val = POP().int_val;
                printf("print: %lld\n", (long long)int_val);
            }
            break;

            default:
            fprintf(stderr, "unknown bytecode instruction\n");
            exit(-1);
        }
    }
    */



}

fn main()
{
    let args: Vec<String> = env::args().collect();
    println!("{:?}", args);

    if args.len() == 2 {
        let prog = parse_file(&args[1]);
        eval(prog);
    }
}
